src/raftstore/store/peer.rs

```diff
@@ -350,6 +355,47 @@ impl Peer {
         Ok(())
     }
 
+    pub fn step(&mut self, m: eraftpb::Message) -> Result<()> {
+        if self.is_leader() && m.get_from() != INVALID_ID {
+            self.peer_heartbeats.insert(m.get_from(), Instant::now());
+        }
+        try!(self.raft_group.step(m));
+        Ok(())
+    }
+
+    pub fn check_peers(&mut self) {
+        let peers = self.get_store().get_region().get_peers().to_owned();
+        for peer in &peers {
+            self.peer_heartbeats.entry(peer.get_id()).or_insert_with(Instant::now);
+        }
+        let removed_ids: Vec<_> = self.peer_heartbeats
+            .keys()
+            .filter(|&&id| !peers.iter().any(|p| p.get_id() == id))
+            .cloned()
+            .collect();
+        for id in removed_ids {
+            self.peer_heartbeats.remove(&id);
+        }
+    }
+
+    pub fn collect_down_peers(&self, max_duration: Duration) -> Vec<PeerStats> {
+        let mut down_peers = Vec::new();
+        for p in self.get_store().get_region().get_peers() {
+            if p.get_id() == self.peer.get_id() {
+                continue;
+            }
+            if let Some(instant) = self.peer_heartbeats.get(&p.get_id()) {
+                if instant.elapsed() >= max_duration {
```

Comment:
If A was a leader, then it steps down as follower. After `max_duration`, it becomes leader again. Will all the followers be treated as down peer?